### PR TITLE
Extending support for setting thread count on I/O.

### LINKED ISSUE #69: Allow setting I/O threads everywhere htslib allows it.

### PR DESCRIPTION
**Pull request by holtgrewe**
fixes #69

### CODE REVIEW

File: src/bam/mod.rs [hunk 1]
```diff
@@ -124,6 +124,29 @@ impl Reader {
...
+    /// reading of large BAM files.
+    ///
+    /// Setting `nthreads` to `0` does not change the current state.  Note that it is not
+    /// possible to set the number of background threads below `1` once it has been set.
+    ///
+    /// # Arguments
+    ///
+    /// * `n_threads` - number of extra background reader threads to use
+    pub fn set_threads(&mut self, n_threads: usize) -> Result<(), ThreadingError> {
+        if n_threads == 0 {
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bam/mod.rs]: Moving.

File: src/bam/mod.rs [hunk 2]
```diff
@@ -388,10 +434,19 @@ impl Writer {
...
     /// # Arguments
     ///
-    /// * `n_threads` - number of background writer threads to use
+    /// * `n_threads` - number of extra background writer threads to use
     pub fn set_threads(&mut self, n_threads: usize) -> Result<(), ThreadingError> {
+        if n_threads == 0 {
+            // Do nothing if thread count is zero.  bgzf_mt() below would
+            // return -1 to indicate an error.
+            return Ok(());
+        }
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bam/mod.rs]: OK

File: src/bcf/mod.rs [hunk 3]
```diff
@@ -96,6 +96,29 @@ impl Reader {
...
+    /// Activate multi-threaded BCF read support in htslib. This should permit faster
+    /// reading of large BCF files.
+    ///
+    /// Setting `nthreads` to `0` does not change the current state.  Note that it is not
+    /// possible to set the number of background threads below `1` once it has been set.
+    ///
+    /// # Arguments
+    ///
+    /// * `n_threads` - number of extra background reader threads to use
+    pub fn set_threads(&mut self, n_threads: usize) -> Result<(), ThreadingError> {
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/mod.rs]: Ok

File: src/bcf/mod.rs [hunk 4]
```diff
@@ -37,6 +37,25 @@ pub struct Reader {
 unsafe impl Send for Reader {}
 
 
+/// Implementation for `Reader::set_threads()` and `Writer::set_threads`.
+pub fn set_threads(hts_file: *mut htslib::htsFile, n_threads: usize)
+        -> Result<(), ThreadingError> {
+    assert!(n_threads > 0, "n_threads must be > 0");
+
+    if n_threads == 0 {
```
Comment (johanneskoester) [inline src/bcf/mod.rs]: Please remove the if since this is already handled by the assertion above.

File: src/tbx/mod.rs [hunk 5]
```diff
@@ -204,6 +204,8 @@ impl Reader {
     ///
     /// * `n_threads` - number of extra background reader threads to use
     pub fn set_threads(&mut self, n_threads: usize) -> Result<(), ThreadingError> {
+        assert!(n_threads > 0, "n_threads must be > 0");
```
Comment (johanneskoester) [inline src/tbx/mod.rs]: Am I right that we can simply use  `bcf::set_threads` here? Maybe move it to utils.rs before.

File: src/bcf/mod.rs (hunk 4 above)
Comment (holtgrewe) [reply src/bcf/mod.rs]: Of course.

File: src/tbx/mod.rs (hunk 5 above)
Comment (holtgrewe) [reply src/tbx/mod.rs]: We would need a common `ThreadingError` type then...

Comment (holtgrewe) [reply src/tbx/mod.rs]: Can we maybe go forward with this patch anyway? I don't know so much about Rust error types and the best practice here (in python it's considered best practice to have an `exceptions` module for a library). Also, I'm planning to add an `IndexedReader` for VCF/BCF soon anyway and we can some more cleanup afterwards and consolidate the shared things? E.g., we should be able to centralize all rust code for parallel bgzf into one place including the thread pool stuff I am planning to add.

Comment (johanneskoester) [reply src/tbx/mod.rs]: Ok!

Comment (johanneskoester) [reply src/tbx/mod.rs]: The most standard solution for error handling is using the error-chain crate. I want to migrate error handling to that, but I did not find the time so far. See https://docs.rs/error-chain

**Pull request merged**